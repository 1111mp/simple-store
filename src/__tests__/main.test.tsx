import { fireEvent, render, waitFor } from "@testing-library/react";

import { createStore } from "../";
import { useEffect, useState } from "react";

type UserStore = {
  name: string;
  age: number;
};

function sleep(duration: number): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve();
    }, duration);
  });
}

describe("test simple-store", () => {
  test("basic store", () => {
    const [userStore] = createStore<UserStore>({
      name: "Tom",
      age: 18,
    });

    const App: React.FC = () => {
      const [{ name, age }, updateUserStore] = userStore();

      return (
        <div>
          App:
          <br />
          name: <span data-testid="name">{name}</span>
          <br />
          age: <span data-testid="age">{age}</span>
          <br />
          <button
            onClick={() => {
              updateUserStore((store) => {
                return { name: "Jim", age: 16 };
              });
            }}
          >
            update
          </button>
        </div>
      );
    };

    const { asFragment, getByTestId, getByText } = render(<App />);

    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId("name")).toHaveTextContent("Tom");
    expect(getByTestId("age")).toHaveTextContent("18");

    fireEvent.click(getByText("update"));

    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId("name")).toHaveTextContent("Jim");
    expect(getByTestId("age")).toHaveTextContent("16");
  });

  test("hooks for use store", () => {
    const [userStore] = createStore<UserStore>({
      name: "Tom",
      age: 18,
    });

    function useUserStore() {
      const [{ name, age }, updateStore] = userStore();

      const updateName = (name: string) => {
        updateStore((store) => ({ ...store, name }));
      };

      return {
        name,
        age,
        updateName,
      };
    }

    const App: React.FC = () => {
      const { name, age, updateName } = useUserStore();

      return (
        <div>
          App:
          <br />
          name: <span data-testid="name">{name}</span>
          <br />
          age: <span data-testid="age">{age}</span>
          <br />
          <button
            onClick={() => {
              updateName("Jim");
            }}
          >
            update
          </button>
        </div>
      );
    };

    const { asFragment, getByTestId, getByText } = render(<App />);

    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId("name")).toHaveTextContent("Tom");
    expect(getByTestId("age")).toHaveTextContent("18");

    fireEvent.click(getByText("update"));

    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId("name")).toHaveTextContent("Jim");
    expect(getByTestId("age")).toHaveTextContent("18");
  });

  test("multiple component", () => {
    const [userStore] = createStore<UserStore>({
      name: "Tom",
      age: 18,
    });

    const Other: React.FC = () => {
      const [{ name, age }] = userStore();

      return (
        <div>
          Other
          <br />
          name: <span data-testid="other-name">{name}</span>
          <br />
          age: <span data-testid="other-age">{age}</span>
          <br />
        </div>
      );
    };

    const Test: React.FC = () => {
      const [{ name, age }, updateStore] = userStore();

      return (
        <div>
          Test
          <br />
          name: <span data-testid="test-name">{name}</span>
          <br />
          age: <span data-testid="test-age">{age}</span>
          <br />
          <button
            onClick={() => {
              updateStore(() => ({ name: "Jim", age: 16 }));
            }}
          >
            update
          </button>
        </div>
      );
    };

    const App: React.FC = () => {
      return (
        <div>
          <Test />
          <br />
          <Other />
        </div>
      );
    };

    const { asFragment, getByTestId, getByText } = render(<App />);

    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId("test-name")).toHaveTextContent("Tom");
    expect(getByTestId("test-age")).toHaveTextContent("18");
    expect(getByTestId("other-name")).toHaveTextContent("Tom");
    expect(getByTestId("other-age")).toHaveTextContent("18");

    fireEvent.click(getByText("update"));

    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId("test-name")).toHaveTextContent("Jim");
    expect(getByTestId("test-age")).toHaveTextContent("16");
    expect(getByTestId("other-name")).toHaveTextContent("Jim");
    expect(getByTestId("other-age")).toHaveTextContent("16");
  });

  test("depsFn", () => {
    const [userStore] = createStore<UserStore>({
      name: "Tom",
      age: 18,
    });

    const App: React.FC = () => {
      const [{ name, age }, updateUserStore] = userStore((store) => [
        store.name,
      ]);

      return (
        <div>
          App:
          <br />
          name: <span data-testid="name">{name}</span>
          <br />
          age: <span data-testid="age">{age}</span>
          <br />
          <button
            onClick={() => {
              updateUserStore((store) => {
                return { ...store, age: 20 };
              });
            }}
          >
            grow up
          </button>
        </div>
      );
    };

    const { asFragment, getByTestId, getByText } = render(<App />);

    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId("name")).toHaveTextContent("Tom");
    expect(getByTestId("age")).toHaveTextContent("18");

    fireEvent.click(getByText("grow up"));

    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId("name")).toHaveTextContent("Tom");
    expect(getByTestId("age")).toHaveTextContent("18");
  });

  test("multiple component with depsFn", () => {
    const [userStore] = createStore<UserStore>({
      name: "Tom",
      age: 18,
    });

    const Other: React.FC = () => {
      const [{ name, age }] = userStore(() => []);

      return (
        <div>
          Other
          <br />
          name: <span data-testid="other-name">{name}</span>
          <br />
          age: <span data-testid="other-age">{age}</span>
          <br />
        </div>
      );
    };

    const Test: React.FC = () => {
      const [{ name, age }, updateStore] = userStore();

      return (
        <div>
          Test
          <br />
          name: <span data-testid="test-name">{name}</span>
          <br />
          age: <span data-testid="test-age">{age}</span>
          <br />
          <button
            onClick={() => {
              updateStore(() => ({ name: "Jim", age: 16 }));
            }}
          >
            update
          </button>
        </div>
      );
    };

    const App: React.FC = () => {
      return (
        <div>
          <Test />
          <br />
          <Other />
        </div>
      );
    };

    const { asFragment, getByTestId, getByText } = render(<App />);

    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId("test-name")).toHaveTextContent("Tom");
    expect(getByTestId("test-age")).toHaveTextContent("18");
    expect(getByTestId("other-name")).toHaveTextContent("Tom");
    expect(getByTestId("other-age")).toHaveTextContent("18");

    fireEvent.click(getByText("update"));

    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId("test-name")).toHaveTextContent("Jim");
    expect(getByTestId("test-age")).toHaveTextContent("16");
    expect(getByTestId("other-name")).toHaveTextContent("Tom");
    expect(getByTestId("other-age")).toHaveTextContent("18");
  });

  test("async update", async () => {
    const [userStore] = createStore<UserStore>({
      name: "Tom",
      age: 18,
    });

    const App: React.FC = () => {
      const [{ name, age }, updateUserStore] = userStore();

      return (
        <div>
          App:
          <br />
          name: <span data-testid="name">{name}</span>
          <br />
          age: <span data-testid="age">{age}</span>
          <br />
          <br />
          <button
            onClick={async () => {
              await sleep(1000);
              updateUserStore((store) => {
                return { name: "Jim", age: 16 };
              });
            }}
          >
            update
          </button>
        </div>
      );
    };

    const { asFragment, getByTestId, getByText } = render(<App />);

    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId("name")).toHaveTextContent("Tom");
    expect(getByTestId("age")).toHaveTextContent("18");

    await waitFor(
      async () => {
        fireEvent.click(getByText("update"));
        await sleep(1500);
      },
      { timeout: 2000 }
    );

    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId("name")).toHaveTextContent("Jim");
    expect(getByTestId("age")).toHaveTextContent("16");
  });

  test("use update promise", async () => {
    const [userStore] = createStore<UserStore>({
      name: "Tom",
      age: 18,
    });

    const App: React.FC = () => {
      const [{ name, age }, updateUserStore] = userStore();

      return (
        <div>
          App:
          <br />
          name: <span data-testid="name">{name}</span>
          <br />
          age: <span data-testid="age">{age}</span>
          <br />
          <br />
          <button
            onClick={() => {
              updateUserStore(async () => {
                await sleep(1000);
                return { name: "Jim", age: 16 };
              });
            }}
          >
            update
          </button>
        </div>
      );
    };

    const { asFragment, getByTestId, getByText } = render(<App />);

    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId("name")).toHaveTextContent("Tom");
    expect(getByTestId("age")).toHaveTextContent("18");

    await waitFor(
      async () => {
        fireEvent.click(getByText("update"));
        await sleep(1500);
      },
      { timeout: 2000 }
    );

    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId("name")).toHaveTextContent("Jim");
    expect(getByTestId("age")).toHaveTextContent("16");
  });

  test("resetStore api", async () => {
    const [userStore, resetUserStore] = createStore<UserStore>({
      name: "Tom",
      age: 18,
    });

    const Test: React.FC = () => {
      const [{ name, age }, updateUserStore] = userStore();

      useEffect(() => {
        return () => {
          resetUserStore();
        };
      }, []);

      return (
        <div>
          Test
          <br />
          name: <span data-testid="test-name">{name}</span>
          <br />
          age: <span data-testid="test-age">{age}</span>
          <br />
          <button
            onClick={() => {
              updateUserStore((store) => ({
                ...store,
                name: "Jim",
                age: 16,
              }));
            }}
          >
            update
          </button>
        </div>
      );
    };

    const App: React.FC = () => {
      const [show, setShow] = useState<boolean>(true);
      const [{ name, age }] = userStore();

      return (
        <div>
          App:
          <br />
          name: <span data-testid="app-name">{name}</span>
          <br />
          age: <span data-testid="app-age">{age}</span>
          <br />
          <button
            onClick={() => {
              setShow(false);
            }}
          >
            hide
          </button>
          <br />
          {show ? <Test /> : null}
        </div>
      );
    };

    const { asFragment, getByTestId, getByText, queryByTestId } = render(
      <App />
    );

    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId("app-name")).toHaveTextContent("Tom");
    expect(getByTestId("app-age")).toHaveTextContent("18");
    expect(getByTestId("test-name")).toHaveTextContent("Tom");
    expect(getByTestId("test-age")).toHaveTextContent("18");

    fireEvent.click(getByText("update"));

    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId("app-name")).toHaveTextContent("Jim");
    expect(getByTestId("app-age")).toHaveTextContent("16");
    expect(getByTestId("test-name")).toHaveTextContent("Jim");
    expect(getByTestId("test-age")).toHaveTextContent("16");

    fireEvent.click(getByText("hide"));

    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId("app-name")).toHaveTextContent("Tom");
    expect(getByTestId("app-age")).toHaveTextContent("18");
    expect(queryByTestId("test-name")).not.toBeInTheDocument();
    expect(queryByTestId("test-age")).not.toBeInTheDocument();
  });
});
