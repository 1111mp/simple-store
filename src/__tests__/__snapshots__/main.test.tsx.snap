// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test simple-store async update 1`] = `
<DocumentFragment>
  <div>
    App:
    <br />
    name: 
    <span
      data-testid="name"
    >
      Tom
    </span>
    <br />
    age: 
    <span
      data-testid="age"
    >
      18
    </span>
    <br />
    <br />
    <button>
      update
    </button>
  </div>
</DocumentFragment>
`;

exports[`test simple-store async update 2`] = `
<DocumentFragment>
  <div>
    App:
    <br />
    name: 
    <span
      data-testid="name"
    >
      Jim
    </span>
    <br />
    age: 
    <span
      data-testid="age"
    >
      16
    </span>
    <br />
    <br />
    <button>
      update
    </button>
  </div>
</DocumentFragment>
`;

exports[`test simple-store basic store 1`] = `
<DocumentFragment>
  <div>
    App:
    <br />
    name: 
    <span
      data-testid="name"
    >
      Tom
    </span>
    <br />
    age: 
    <span
      data-testid="age"
    >
      18
    </span>
    <br />
    <button>
      update
    </button>
  </div>
</DocumentFragment>
`;

exports[`test simple-store basic store 2`] = `
<DocumentFragment>
  <div>
    App:
    <br />
    name: 
    <span
      data-testid="name"
    >
      Jim
    </span>
    <br />
    age: 
    <span
      data-testid="age"
    >
      16
    </span>
    <br />
    <button>
      update
    </button>
  </div>
</DocumentFragment>
`;

exports[`test simple-store depsFn 1`] = `
<DocumentFragment>
  <div>
    App:
    <br />
    name: 
    <span
      data-testid="name"
    >
      Tom
    </span>
    <br />
    age: 
    <span
      data-testid="age"
    >
      18
    </span>
    <br />
    <button>
      grow up
    </button>
  </div>
</DocumentFragment>
`;

exports[`test simple-store depsFn 2`] = `
<DocumentFragment>
  <div>
    App:
    <br />
    name: 
    <span
      data-testid="name"
    >
      Tom
    </span>
    <br />
    age: 
    <span
      data-testid="age"
    >
      18
    </span>
    <br />
    <button>
      grow up
    </button>
  </div>
</DocumentFragment>
`;

exports[`test simple-store hooks for use store 1`] = `
<DocumentFragment>
  <div>
    App:
    <br />
    name: 
    <span
      data-testid="name"
    >
      Tom
    </span>
    <br />
    age: 
    <span
      data-testid="age"
    >
      18
    </span>
    <br />
    <button>
      update
    </button>
  </div>
</DocumentFragment>
`;

exports[`test simple-store hooks for use store 2`] = `
<DocumentFragment>
  <div>
    App:
    <br />
    name: 
    <span
      data-testid="name"
    >
      Jim
    </span>
    <br />
    age: 
    <span
      data-testid="age"
    >
      18
    </span>
    <br />
    <button>
      update
    </button>
  </div>
</DocumentFragment>
`;

exports[`test simple-store multiple component 1`] = `
<DocumentFragment>
  <div>
    <div>
      Test
      <br />
      name: 
      <span
        data-testid="test-name"
      >
        Tom
      </span>
      <br />
      age: 
      <span
        data-testid="test-age"
      >
        18
      </span>
      <br />
      <button>
        update
      </button>
    </div>
    <br />
    <div>
      Other
      <br />
      name: 
      <span
        data-testid="other-name"
      >
        Tom
      </span>
      <br />
      age: 
      <span
        data-testid="other-age"
      >
        18
      </span>
      <br />
    </div>
  </div>
</DocumentFragment>
`;

exports[`test simple-store multiple component 2`] = `
<DocumentFragment>
  <div>
    <div>
      Test
      <br />
      name: 
      <span
        data-testid="test-name"
      >
        Jim
      </span>
      <br />
      age: 
      <span
        data-testid="test-age"
      >
        16
      </span>
      <br />
      <button>
        update
      </button>
    </div>
    <br />
    <div>
      Other
      <br />
      name: 
      <span
        data-testid="other-name"
      >
        Jim
      </span>
      <br />
      age: 
      <span
        data-testid="other-age"
      >
        16
      </span>
      <br />
    </div>
  </div>
</DocumentFragment>
`;

exports[`test simple-store multiple component with depsFn 1`] = `
<DocumentFragment>
  <div>
    <div>
      Test
      <br />
      name: 
      <span
        data-testid="test-name"
      >
        Tom
      </span>
      <br />
      age: 
      <span
        data-testid="test-age"
      >
        18
      </span>
      <br />
      <button>
        update
      </button>
    </div>
    <br />
    <div>
      Other
      <br />
      name: 
      <span
        data-testid="other-name"
      >
        Tom
      </span>
      <br />
      age: 
      <span
        data-testid="other-age"
      >
        18
      </span>
      <br />
    </div>
  </div>
</DocumentFragment>
`;

exports[`test simple-store multiple component with depsFn 2`] = `
<DocumentFragment>
  <div>
    <div>
      Test
      <br />
      name: 
      <span
        data-testid="test-name"
      >
        Jim
      </span>
      <br />
      age: 
      <span
        data-testid="test-age"
      >
        16
      </span>
      <br />
      <button>
        update
      </button>
    </div>
    <br />
    <div>
      Other
      <br />
      name: 
      <span
        data-testid="other-name"
      >
        Tom
      </span>
      <br />
      age: 
      <span
        data-testid="other-age"
      >
        18
      </span>
      <br />
    </div>
  </div>
</DocumentFragment>
`;

exports[`test simple-store resetStore api 1`] = `
<DocumentFragment>
  <div>
    App:
    <br />
    name: 
    <span
      data-testid="app-name"
    >
      Tom
    </span>
    <br />
    age: 
    <span
      data-testid="app-age"
    >
      18
    </span>
    <br />
    <button>
      hide
    </button>
    <br />
    <div>
      Test
      <br />
      name: 
      <span
        data-testid="test-name"
      >
        Tom
      </span>
      <br />
      age: 
      <span
        data-testid="test-age"
      >
        18
      </span>
      <br />
      <button>
        update
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`test simple-store resetStore api 2`] = `
<DocumentFragment>
  <div>
    App:
    <br />
    name: 
    <span
      data-testid="app-name"
    >
      Jim
    </span>
    <br />
    age: 
    <span
      data-testid="app-age"
    >
      16
    </span>
    <br />
    <button>
      hide
    </button>
    <br />
    <div>
      Test
      <br />
      name: 
      <span
        data-testid="test-name"
      >
        Jim
      </span>
      <br />
      age: 
      <span
        data-testid="test-age"
      >
        16
      </span>
      <br />
      <button>
        update
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`test simple-store resetStore api 3`] = `
<DocumentFragment>
  <div>
    App:
    <br />
    name: 
    <span
      data-testid="app-name"
    >
      Tom
    </span>
    <br />
    age: 
    <span
      data-testid="app-age"
    >
      18
    </span>
    <br />
    <button>
      hide
    </button>
    <br />
  </div>
</DocumentFragment>
`;

exports[`test simple-store use update promise 1`] = `
<DocumentFragment>
  <div>
    App:
    <br />
    name: 
    <span
      data-testid="name"
    >
      Tom
    </span>
    <br />
    age: 
    <span
      data-testid="age"
    >
      18
    </span>
    <br />
    <br />
    <button>
      update
    </button>
  </div>
</DocumentFragment>
`;

exports[`test simple-store use update promise 2`] = `
<DocumentFragment>
  <div>
    App:
    <br />
    name: 
    <span
      data-testid="name"
    >
      Jim
    </span>
    <br />
    age: 
    <span
      data-testid="age"
    >
      16
    </span>
    <br />
    <br />
    <button>
      update
    </button>
  </div>
</DocumentFragment>
`;
